File: backend/app/models/asset.py
"""Asset metadata model — registered file record (does not execute cues)."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING, Any
from uuid import uuid4

from sqlalchemy import JSON, BigInteger, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base
from app.models.tag import asset_tags

if TYPE_CHECKING:
    from app.models.tag import Tag


class AssetType(StrEnum):
    VIDEO = "video"
    AUDIO = "audio"
    IMAGE = "image"
    TEXT = "text"
    DOCUMENT = "document"
    DATA = "data"
    OTHER = "other"


ASSET_TYPES = frozenset(t.value for t in AssetType)


class Asset(Base):
    """Metadata for a file belonging to a production.

    Does not store binary content and does not execute cues.
    ``metadata_json`` maps to the DB/API field ``metadata`` (SQLAlchemy reserves
    ``Base.metadata``).
    """

    __tablename__ = "assets"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid4()))
    production_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("productions.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    type: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    original_filename: Mapped[str] = mapped_column(String(500), nullable=False)
    storage_key: Mapped[str] = mapped_column(String(512), nullable=False)
    mime_type: Mapped[str] = mapped_column(String(200), nullable=False)
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    checksum: Mapped[str] = mapped_column(String(128), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    # Column name "metadata"; attr must not collide with Declarative Base.metadata.
    metadata_json: Mapped[dict[str, Any]] = mapped_column(
        "metadata",
        JSON,
        nullable=False,
        default=dict,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    tags: Mapped[list[Tag]] = relationship(
        "Tag",
        secondary=asset_tags,
        back_populates="assets",
        order_by="Tag.name",
    )

File: frontend/lib/types/asset.ts
import type { Tag } from "@/lib/types/tag";

export type AssetType =
  | "video"
  | "audio"
  | "image"
  | "text"
  | "document"
  | "data"
  | "other";

export const ASSET_TYPES: AssetType[] = [
  "video",
  "audio",
  "image",
  "text",
  "document",
  "data",
  "other"
];

export type Asset = {
  id: string;
  production_id: string;
  name: string;
  type: AssetType;
  original_filename: string;
  storage_key: string;
  mime_type: string;
  size_bytes: number;
  checksum: string;
  description: string | null;
  metadata: Record<string, unknown>;
  tags: Tag[];
  created_at: string;
  updated_at: string;
};

export type AssetPreview = {
  asset_id: string;
  production_id: string;
  mime_type: string;
  kind: "image" | "audio" | "video" | "text" | "json" | "none";
  preview_available: boolean;
  text_excerpt: string | null;
  size_bytes: number;
  original_filename: string;
};

export type AssetUpdateInput = {
  name?: string;
  type?: AssetType;
  description?: string | null;
  metadata?: Record<string, unknown>;
};
